### PR TITLE
Flink runner: refactor the translator into two phases: rewriting and translating.

### PR DESCRIPTION
**Pull request by peihe**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`.
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

Background of this PR:
I need PipelineTranslationOptimizer to annotate the perTransformParallelism before the translator.
However, because translator also do the rewriting, the optimizer cannot annotate the whole graph.
Separating the rewriting from the translating would make it possible to perform additional phases in-between.

Changes that leads to this PR:
https://github.com/apache/beam/compare/master...peihe:configuration-1?expand=1

### CODE REVIEW

**Comment by peihe** [comment]
R: @tgroh @mxm @aljoscha

File: runners/flink/src/main/java/org/apache/beam/runners/flink/PipelineTranslationRewriter.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    this.isStreaming = flinkRunner.getPipelineOptions().isStreaming();
+  }
+
+  @Override
+  public void translate(Pipeline pipeline) {
+    pipeline.replaceAll(getOverrides());
+  }
+
+  private List<PTransformOverride> getOverrides() {
+    if (isStreaming) {
```
Comment (tgroh) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/PipelineTranslationRewriter.java]: Can this not just be inlined in the appropriate spot? We can have a utility class for the overrides if we want, but I don't think it's hugely useful, and we can call the pipeline.replaceAll(overrides) right before trying to translate via the streaming and/or batch pipeline translator. I don't think this additional abstraction layer gets us anything.

Comment (peihe) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/PipelineTranslationRewriter.java]: Is the suggestion to:
1. remove getOverrides(), and inline if(isStreaming) {} else {} in translate() {} above.
I can do this.

2. or to create separate rewriters for batch and streaming.
I prefer to have one class works for both batch and streaming.

3. or to drop this PR? 
As I stated in the PR description, I need this PR to separate the phase of annotation from the translating (plus rewriting).

**Comment by aljoscha** [comment]
@peihe I have #2915 open, which changes the translation process a bit and introduces better separation. (As a side effect of introducing asynchronous job submission.)

**Comment by peihe** [comment]
@aljoscha 
Looking at #2915, I think two PRs are orthogonal (few lines conflicts in FlinkPipelineExecutionEnvironment).

I think we can get this PR reviewed, and then we can sync about the order of merging.

**Comment by aljoscha** [comment]
Agreed, let's see what @tgroh responds to your comments.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/PipelineTranslationRewriter.java (hunk 1 above)
Comment (tgroh) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/PipelineTranslationRewriter.java]: mostly 1 and a bit of 2. I would prefer to split the `streamingTransformOverrides()` and `batchTransformOverrides()` into separate methods just for readability.

More importantly, the idea would be to remove the name `PipelineTranslationRewriter`, and just make it a `FlinkTransformOverrides` class that, given a streaming/batch configuration returns the list of `PTransformOverride` that is appropriate for the mode (or alternatively just exposes `streamingTransformOverrides()` and `batchTransformOverrides()` methods). Then just call `pipeline.replaceAll(overrides)` in the call to `run(Pipeline)`. I don't believe the `rewriter` abstraction to be useful.

**Comment by peihe** [comment]
PTAL @tgroh

**Comment by tgroh** [comment]
I don't see any changes since my last comment

**Comment by peihe** [comment]
Updated and resolved conflicts.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkTransformOverrides.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    if (streaming) {
+      return ImmutableList.<PTransformOverride>builder()
+          .add(
+              PTransformOverride.of(
+                  PTransformMatchers.splittableParDoMulti(),
+                  new FlinkStreamingPipelineTranslator.SplittableParDoOverrideFactory()))
+          .add(
+              PTransformOverride.of(
+                  PTransformMatchers.classEqualTo(SplittableParDo.ProcessKeyedElements.class),
+                  new SplittableParDoViaKeyedWorkItems.OverrideFactory()))
```
Comment (tgroh) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkTransformOverrides.java]: These factories already existed? Cool.

**Pull request closed without merging**